workflow/docs/why-do-we-not-use-argo.md

```diff
@@ -0,0 +1,49 @@
+# Why do we not use Argo
+<!-- TOC -->
+
+- [Why do we not use Argo](#why-do-we-not-use-argo)
+  - [What's the advantage if we use argo](#whats-the-advantage-if-we-use-argo)
+  - [Argo workflow is not the best way to describe declarative chaos experiments](#argo-workflow-is-not-the-best-way-to-describe-declarative-chaos-experiments)
+  - [We could do more things if we have a customized workflow](#we-could-do-more-things-if-we-have-a-customized-workflow)
+  - [Why the other chaos platform choose to use Argo, but we do not prefer to use it](#why-the-other-chaos-platform-choose-to-use-argo-but-we-do-not-prefer-to-use-it)
+  - [Other misc](#other-misc)
+
+<!-- /TOC -->
+
+It has been a long time that we decide to build up another workflow engine. I believe that we have enough reasons that push us to build a new workflow engine.
+
+But when other contributors ask me, "Why not use Argo"? I could not give them convincing words in time. That's why I decide to write this doc.
+
+## What's the advantage if we use argo
+
+Argo is an amazing workflow engine for common purpose. Argo provides "sequential/parallel execution", "conditional branches", "DAG"; and workloads like "script", "container", "resource".
+
+I think the most advantage is reducing our jobs, we do not need to re-implement another workflow engine. We could build a workflow by writing some scripts with `kubectl`.
+
+## Argo workflow is not the best way to describe declarative chaos experiments
+
+Argo supports "script", "container" and "resource" workloads, it could integrate with Chaos Mesh to apply/recover chaos experiments. But this way to apply chaos experiments is not declarative anymore, it turns to imperative.
+
+Argo is so powerful and it's impossible to ask users always for following the best practices. Users could write complex scripts for determining the situation, then `kubectl apply` or `kubectl patch` something, once a command make unexpected side-effects, it will let whole system unstable.
+
+We define the workflow that could describe chaos experiments in the declarative, changing chaos experiments is just like "from expected status A turning into expected status B", users should not considering about "remove some chaos CR object" or "create new chaos CR object". We also allow users (only) to execute their scripts in `Task` node, it will not affect the changing of chaos experiment states.
```

Comment:
Is there now any thoughts on a declarative way of defining workflow status? I don't mean to judge, just being curious. You see, I suppose there has to be somewhere to store the status, and we all agree that using a CRD is an idea not that good, not being native to k8s. If possible, I'm looking forward to a further and deeper discussion or explaination on the designing of workflow.

Reply:
> You see, I suppose there has to be somewhere to store the status, and we all agree that using a CRD is an idea not that good, not being native to k8s.

I think using CRD could not be determined as "good" or "bad", CRD is the only way that we store the status by Kubernetes API.

I want to present that "state machine with event-driven" is bad. It has some discussions at https://github.com/chaos-mesh/chaos-mesh/pull/1418#issuecomment-765224545

And I am still trying to implement workflow without state-machine anymore, it is nearly accomplished (on theory). The core is defining the expected status more clearly, makes as more as possible thing have been determined as the object be created.

I will update the design documents after I finish the proving of concepts.